File: cmd/agent/registration.py
"""Machine registration client for Beta9 gateway.

Handles POST /api/v1/machine/register to transition machine from
pending → registered status.
"""

import logging
from dataclasses import dataclass
from typing import Optional

import httpx

from .config import AgentConfig
from .metrics import get_cpu_string, get_memory_string, get_private_ip, detect_gpu_count

logger = logging.getLogger("beta9-agent")


@dataclass
class RegistrationResult:
    """Result of machine registration."""
    success: bool
    config: Optional[dict] = None
    error: Optional[str] = None


def register_machine(config: AgentConfig) -> RegistrationResult:
    """Register machine with Beta9 gateway.

    Sends POST /api/v1/machine/register with machine details.

    Args:
        config: Agent configuration

    Returns:
        RegistrationResult with success status and gateway config
    """
    hostname = f"machine-{config.machine_id}"

    payload = {
        "token": config.k3s_token or "mock-k3s-token",  # k3s token, not registration token
        "machine_id": config.machine_id,
        "hostname": hostname,
        "provider_name": config.provider_name,
        "pool_name": config.pool_name,
        "cpu": get_cpu_string(),
        "memory": get_memory_string(),
        "gpu_count": str(detect_gpu_count()),
        "private_ip": get_private_ip(),
    }

    logger.info(f"Registering machine {config.machine_id} with gateway {config.gateway_url}")
    logger.debug(f"Registration payload: {payload}")

    if config.dry_run:
        logger.info("Dry run mode - skipping actual registration")
        return RegistrationResult(success=True, config={"dry_run": True})

    try:
        headers = {
            "Authorization": f"Bearer {config.token}",
            "Content-Type": "application/json",
        }

        with httpx.Client(timeout=config.registration_timeout) as client:
            response = client.post(
                config.register_url,
                json=payload,
                headers=headers,
            )

            if response.status_code == 200:
                data = response.json()
                logger.info(f"Machine {config.machine_id} registered successfully")
                return RegistrationResult(
                    success=True,
                    config=data.get("config"),
                )

            elif response.status_code == 403:
                return RegistrationResult(
                    success=False,
                    error="Invalid token - ensure token is from 'beta9 machine create'",
                )

            elif response.status_code == 400:
                return RegistrationResult(
                    success=False,
                    error=f"Bad request: {response.text}",
                )

            else:
                return RegistrationResult(
                    success=False,
                    error=f"Unexpected status {response.status_code}: {response.text}",
                )

    except httpx.ConnectError as e:
        return RegistrationResult(
            success=False,
            error=f"Connection failed to {config.gateway_url}: {e}",
        )

    except httpx.TimeoutException:
        return RegistrationResult(
            success=False,
            error=f"Timeout connecting to {config.gateway_url}",
        )

    except Exception as e:
        return RegistrationResult(
            success=False,
            error=f"Registration failed: {e}",
        )

File: pkg/agent/registration.go
package agent

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strconv"

	"github.com/rs/zerolog/log"
)

// RegistrationPayload is the request body for machine registration
type RegistrationPayload struct {
	Token        string `json:"token"`
	MachineID    string `json:"machine_id"`
	Hostname     string `json:"hostname"`
	ProviderName string `json:"provider_name"`
	PoolName     string `json:"pool_name"`
	CPU          string `json:"cpu"`
	Memory       string `json:"memory"`
	GPUCount     string `json:"gpu_count"`
	PrivateIP    string `json:"private_ip"`
}

// RegistrationResponse is the response from machine registration
type RegistrationResponse struct {
	Config map[string]interface{} `json:"config"`
}

// RegistrationResult contains the response from registration
type RegistrationResult struct {
	Success bool
	Config  map[string]interface{}
	Error   error
}

// RegisterMachine registers the machine with the Beta9 gateway
func RegisterMachine(ctx context.Context, config *AgentConfig) *RegistrationResult {
	hostname := config.Hostname
	if hostname == "" {
		hostname = fmt.Sprintf("machine-%s", config.MachineID)
	}

	k3sToken := config.K3sToken
	if k3sToken == "" {
		log.Warn().Msg("No --k3s-token provided. Gateway won't be able to deploy worker pods to this machine.")
	}

	gpuCount := DetectGPUCount()

	payload := &RegistrationPayload{
		Token:        k3sToken,
		MachineID:    config.MachineID,
		Hostname:     hostname,
		ProviderName: config.ProviderName,
		PoolName:     config.PoolName,
		CPU:          GetCPUString(),
		Memory:       GetMemoryString(),
		GPUCount:     strconv.Itoa(gpuCount),
		PrivateIP:    GetPrivateIP(),
	}

	log.Info().
		Str("machine_id", config.MachineID).
		Str("gateway", config.GatewayURL()).
		Str("pool", config.PoolName).
		Str("hostname", hostname).
		Int("gpu_count", gpuCount).
		Msg("Registering machine with gateway")

	log.Debug().Interface("payload", payload).Msg("Registration payload")

	if config.DryRun {
		log.Info().Msg("Dry run mode - skipping actual registration")
		return &RegistrationResult{
			Success: true,
			Config:  map[string]interface{}{"dry_run": true},
		}
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return &RegistrationResult{Error: fmt.Errorf("failed to marshal payload: %w", err)}
	}

	client := &http.Client{Timeout: config.RegistrationTimeout}
	req, err := http.NewRequestWithContext(ctx, "POST", config.RegisterURL(), bytes.NewReader(body))
	if err != nil {
		return &RegistrationResult{Error: fmt.Errorf("failed to create request: %w", err)}
	}

	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", config.Token))
	req.Header.Set("Content-Type", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		return &RegistrationResult{
			Error: &ErrRegistrationFailed{
				Reason: fmt.Sprintf("connection failed to %s: %v (is SSH tunnel running?)", config.GatewayURL(), err),
			},
		}
	}
	defer resp.Body.Close()

	respBody, _ := io.ReadAll(resp.Body)

	switch resp.StatusCode {
	case http.StatusOK:
		var result RegistrationResponse
		if err := json.Unmarshal(respBody, &result); err != nil {
			log.Debug().Err(err).Str("body", string(respBody)).Msg("Failed to parse registration response")
		}
		log.Info().
			Str("machine_id", config.MachineID).
			Msg("Machine registered successfully")
		return &RegistrationResult{
			Success: true,
			Config:  result.Config,
		}

	case http.StatusForbidden:
		return &RegistrationResult{
			Error: &ErrRegistrationFailed{
				StatusCode: http.StatusForbidden,
				Reason:     "Invalid token - ensure token is from 'beta9 machine create'",
			},
		}

	case http.StatusBadRequest:
		return &RegistrationResult{
			Error: &ErrRegistrationFailed{
				StatusCode: http.StatusBadRequest,
				Reason:     fmt.Sprintf("Bad request: %s", string(respBody)),
			},
		}

	default:
		return &RegistrationResult{
			Error: &ErrRegistrationFailed{
				StatusCode: resp.StatusCode,
				Reason:     fmt.Sprintf("Unexpected response: %s", string(respBody)),
			},
		}
	}
}
